lib/trino-filesystem/src/test/java/io/trino/filesystem/memory/TestMemoryFileSystemCache.java

```diff
@@ -0,0 +1,131 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.memory;
+
+import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import io.trino.filesystem.AbstractTestTrinoFileSystem;
+import io.trino.filesystem.Location;
+import io.trino.filesystem.TrinoFileSystem;
+import io.trino.filesystem.TrinoInput;
+import io.trino.filesystem.TrinoInputFile;
+import io.trino.filesystem.cache.CacheFileSystem;
+import io.trino.filesystem.cache.CacheKeyProvider;
+import io.trino.filesystem.cache.DefaultCacheKeyProvider;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.UUID;
+
+import static io.airlift.tracing.Tracing.noopTracer;
+import static java.util.concurrent.TimeUnit.HOURS;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestMemoryFileSystemCache
+        extends AbstractTestTrinoFileSystem
+{
+    private static final int MAX_CONTENT_LENGTH = 2 * 1024 * 1024;
+
+    private MemoryFileSystem delegate;
+    private CacheFileSystem fileSystem;
+    private MemoryFileSystemCache cache;
+    private CacheKeyProvider cacheKeyProvider;
+
+    @BeforeAll
+    void beforeAll()
+    {
+        MemoryFileSystemCacheConfig configuration = new MemoryFileSystemCacheConfig()
+                .setMaxContentLength(DataSize.ofBytes(MAX_CONTENT_LENGTH))
+                .setCacheTTL(new Duration(8, HOURS));
+        delegate = new MemoryFileSystem();
+        cache = new MemoryFileSystemCache(noopTracer(), configuration);
+        cacheKeyProvider = new DefaultCacheKeyProvider();
+        fileSystem = new CacheFileSystem(delegate, cache, cacheKeyProvider);
+    }
+
+    @Override
+    protected boolean isHierarchical()
+    {
+        return false;
+    }
+
+    @Override
+    protected boolean supportsCreateExclusive()
+    {
+        return true;
+    }
+
+    @Override
+    protected TrinoFileSystem getFileSystem()
+    {
+        return fileSystem;
+    }
+
+    @Override
+    protected Location getRootLocation()
+    {
+        return Location.of("memory://");
+    }
+
+    @Override
+    protected void verifyFileSystemIsEmpty()
+    {
+        assertThat(delegate.isEmpty()).isTrue();
+    }
+
+    @Test
+    public void testMaxContentLength()
+            throws IOException
+    {
+        int fileSize = MAX_CONTENT_LENGTH + 200;
+        Location location = writeFile(fileSize);
+        TrinoInputFile inputFile = getFileSystem().newInputFile(location);
+        long largeFileSkippedCount = cache.getLargeFileSkippedCount();
+        try (TrinoInput input = inputFile.newInput()) {
+            input.readTail(fileSize);
+        }
+        assertThat(cache.getLargeFileSkippedCount()).isGreaterThan(largeFileSkippedCount);
```

Comment:
nit: Could this assertion be exact?

Reply:
can't make it exact because the test may have parallel accesses to the cache

Reply:
Makes sense. Perhaps we could have a pattern which creates a span and asserts that a particular child span is entered exactly once. Not worth doing here though.